### PR TITLE
CUDA Dispatcher refactor

### LINKED ISSUE #7739: Refactor CUDA target module and conceptual layout to match CPU
The CUDA target implementation shares a lot of the same concepts as the CPU target. It would help with maintenance and implementation in general if the CUDA target were refactored such that the modules (and to some extent classes) match the same layout and concepts as the CPU target.

### PR DESCRIPTION
**Pull request by gmarkall**
This PR moves classes around in the CUDA target to better mirror the structure of the CPU target and core of Numba (see #7739)

Fixes #7739.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/compiler.py [hunk 1]
```diff
@@ -214,14 +191,6 @@ def compile_cuda(pyfunc, return_type, args, debug=False, lineinfo=False,
     return cres
 
 
-def compile_kernel(pyfunc, args, link, debug=False, lineinfo=False,
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py:217]: NOTE: it was determined that this was unused, so is removed.

File: numba/cuda/compiler.py [hunk 2]
```diff
@@ -313,890 +282,3 @@ class ExternFunction(object):
     def __init__(self, name, sig):
         self.name = name
         self.sig = sig
-
-
-class ForAll(object):
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py:318]: Note: `ForAll` was considered a configuration like object so moved to `dispatcher.py`

File: numba/cuda/compiler.py [hunk 3]
```diff
@@ -313,890 +282,3 @@ class ExternFunction(object):
...
-                func=defn._codelibrary.get_cufunc(),
-                b2d_func=0,     # dynamic-shared memory is constant to blksz
-                memsize=self.sharedmem,
-                blocksizelimit=1024,
-            )
-            _, tpb = ctx.get_max_potential_block_size(**kwargs)
-            return tpb
-
-
-class _Kernel(serialize.ReduceMixin):
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py:363]: Note: moved to `dispatcher.py`

File: numba/cuda/compiler.py [hunk 4]
```diff
@@ -313,890 +282,3 @@ class ExternFunction(object):
...
-        elif isinstance(ty, types.BaseTuple):
-            assert len(ty) == len(val)
-            for t, v in zip(ty, val):
-                self._prepare_args(t, v, stream, retr, kernelargs)
-
-        else:
-            raise NotImplementedError(ty, val)
-
-
-class _KernelConfiguration:
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py:709]: Note: moved to `dispatcher.py`

File: numba/cuda/compiler.py [hunk 5]
```diff
@@ -313,890 +282,3 @@ class ExternFunction(object):
...
-                       "to low occupancy.")
-                msg = msg.format(grid=grid_size, sm=2 * smcount)
-                warn(NumbaPerformanceWarning(msg))
-
-    def __call__(self, *args):
-        return self.dispatcher.call(args, self.griddim, self.blockdim,
-                                    self.stream, self.sharedmem)
-
-
-class Dispatcher(_dispatcher.Dispatcher, serialize.ReduceMixin):
```
Comment (stuartarchibald) [inline numba/cuda/compiler.py:733]: Note: moved to `dispatcher.py`

File: numba/cuda/vectorizers.py [hunk 6]
```diff
@@ -1,6 +1,196 @@
 from numba import cuda
-from numba.cuda import dispatcher
+from numpy import array as np_array
 from numba.np.ufunc import deviceufunc
+from numba.np.ufunc.deviceufunc import (UFuncMechanism, GenerializedUFunc,
+                                        GUFuncCallSteps)
+
+
+class CUDAUFuncDispatcher(object):
```
Comment (stuartarchibald) [inline numba/cuda/vectorizers.py:8]: Note: moved from `dispatcher.py` to here, keeping the ufunc logic together.

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/dispatcher.py [hunk 7]
```diff
@@ -1,193 +1,919 @@
...
+        class device_function_template(AbstractTemplate):
+            key = dispatcher
+
+            def generic(self, args, kws):
+                assert not kws
+                return dispatcher.compile(args).signature
+
+            def get_template_info(cls):
+                basepath = os.path.dirname(
+                    os.path.dirname(os.path.dirname(cuda.__file__)))
```
Comment (stuartarchibald) [inline numba/cuda/dispatcher.py:543]: NOTE: Changed to avoid import of `numba` itself.

**Review by stuartarchibald — DISMISSED** [review]
Thanks for doing this refactor, looks good on inspection.

**Comment by gmarkall** [comment]
gpuci run tests

**Review by stuartarchibald — APPROVED** [review]
Re-approved following doc change. Thanks for the patch.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_115`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_115`.

Passed.

**Pull request merged**